node/src/store/utils.rs

```diff
@@ -0,0 +1,548 @@
+use std::fmt::Display;
+use std::iter::once;
+use std::ops::RangeInclusive;
+
+use celestia_types::ExtendedHeader;
+use itertools::Itertools;
+use serde::Serialize;
+use smallvec::{IntoIter, SmallVec};
+use thiserror::Error;
+
+use crate::executor::yield_now;
+use crate::store::{Result, StoreError};
+
+pub(crate) const VALIDATIONS_PER_YIELD: usize = 4;
+
+pub type HeaderRange = RangeInclusive<u64>;
+
+pub(crate) trait RangeLengthExt {
+    fn len(&self) -> u64;
+}
+
+impl RangeLengthExt for RangeInclusive<u64> {
+    fn len(&self) -> u64 {
+        self.end() - self.start() + 1
+    }
+}
```

Comment:
In essence yes, but not exactly - `count` consumes the iterator, since it gets the count by repeatedly calling `next` on it. For `RangeInclusive`, the bounds are known, so we can get the number of elements with maths and avoid clone.
Initially I was hoping to use [ExactSizeIterator](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html), but that's [not implemented for `u64`](https://github.com/rust-lang/rust/issues/36386).

Reply:
`count` can be a footgun because it can overflow... but here we can underflow, could you switch to `saturating_sub()`?

Reply:
good call, switched to `saturating_sub`